src/main/java/com/tabnine/statusBar/StatusBarUpdater.kt

```diff
@@ -0,0 +1,94 @@
+package com.tabnine.statusBar
+
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.project.ProjectManager
+import com.intellij.openapi.wm.WindowManager
+import com.tabnine.binary.BinaryRequestFacade
+import com.tabnine.binary.requests.statusBar.StatusBarPromotionBinaryRequest
+import com.tabnine.binary.requests.statusBar.StatusBarPromotionBinaryResponse
+import com.tabnine.binary.requests.statusBar.StatusBarPromotionShownRequest
+import java.util.concurrent.atomic.AtomicBoolean
+
+class StatusBarUpdater(private val binaryRequestFacade: BinaryRequestFacade) {
+
+    private companion object {
+        val synchronizer = AtomicBoolean()
+    }
+
+    fun updateStatusBar() {
+        ApplicationManager.getApplication().executeOnPooledThread { requestStatusBarMessage() }
+    }
+
+    fun requestStatusBarMessage() {
+        try {
+            if (!synchronizer.compareAndSet(false, true)) {
+                return // abort if there's already a request in progress
+            }
+            val statusBarPromotionWidgets = getStatusBarsWidgets()
+            if (statusBarPromotionWidgets.isEmpty()) {
+                return
+            }
+
+            val promotion = binaryRequestFacade.executeRequest(StatusBarPromotionBinaryRequest())
+
+            if (promotion != null) {
+                updateStatusBars(statusBarPromotionWidgets, promotion)
+                binaryRequestFacade.executeRequest(
+                    StatusBarPromotionShownRequest(
+                        promotion.id,
+                        promotion.message ?: "undefined", promotion.notificationType,
+                        promotion.state
+                    )
+                )
+            } else {
+                clear(statusBarPromotionWidgets)
+            }
+        } finally {
+            synchronizer.set(false)
+        }
+    }
+
+    private fun updateStatusBars(
+        statusBarPromotionWidgets: List<StatusBarPromotionWidget.StatusBarPromotionComponent>,
+        statusBarPromotionResponse: StatusBarPromotionBinaryResponse
+    ) {
+        updateStatusBars(
+            statusBarPromotionWidgets, true,
+            statusBarPromotionResponse.message,
+            statusBarPromotionResponse.id, statusBarPromotionResponse.actions,
+            statusBarPromotionResponse.notificationType
+        )
+    }
+
+    private fun clear(statusBarPromotionWidgets: List<StatusBarPromotionWidget.StatusBarPromotionComponent>) {
+        updateStatusBars(
+            statusBarPromotionWidgets, false, null,
+            null, null, null
+        )
+    }
+
+    private fun updateStatusBars(
```

Comment:
This function gets too many arguments, please refactor it so it will get less (create an object that contains many of the arguments this function get). A good object will probably make line 64-67 much more readable (as you can replace the false and null values with a single null value).